python/dgl/data/csv_dataset.py

```diff
@@ -0,0 +1,420 @@
+                idx = n_data.graph_id == graph_id
+                ids = n_data.id[idx]
+                u_ids, u_indices = np.unique(ids, return_index=True)
+                if len(ids) > len(u_ids):
+                    dgl_warning(
+                        "There exist duplicated ids and only the first ones are kept.")
+                ndata = {}
+                ndata['mapping'] = {index: i for i,
+                                    index in enumerate(ids[u_indices])}
+                data = {}
+                for key, value in n_data.data.items():
+                    data[key] = F.tensor(value[idx][u_indices])
+                ndata['data'] = data
+                if graph_id not in node_dict:
+                    node_dict[graph_id] = {}
+                node_dict[graph_id][n_data.type] = ndata
+        return node_dict
+
+    @staticmethod
+    def _parse_edge_data(edge_data, node_dict):
+        edge_dict = {}
+        for e_data in edge_data:
+            (src_type, e_type, dst_type) = e_data.type
+            graph_ids = np.unique(e_data.graph_id)
+            for graph_id in graph_ids:
+                if graph_id in edge_dict and e_data.type in edge_dict[graph_id]:
+                    raise DGLError(f"Duplicate edge type[{e_data.type}] for same graph[{graph_id}], please place the same edge_type for same graph into single EdgeData.")
+                idx = e_data.graph_id == graph_id
+                edata = {}
+                src_mapping = node_dict[graph_id][src_type]['mapping']
+                dst_mapping = node_dict[graph_id][dst_type]['mapping']
+                src_ids = [src_mapping[index] for index in e_data.src[idx]]
+                dst_ids = [dst_mapping[index] for index in e_data.dst[idx]]
+                edata['edges'] = (
+                    F.tensor(src_ids), F.tensor(dst_ids))
+                data = {}
+                for key, value in e_data.data.items():
+                    data[key] = F.tensor(value[idx])
+                edata['data'] = data
+                if graph_id not in edge_dict:
+                    edge_dict[graph_id] = {}
+                edge_dict[graph_id][e_data.type] = edata
+        return edge_dict
+
+    @staticmethod
+    def _parse_graph_data(graph_data, graphs_dict):
+        if len(graphs_dict) > len(graph_data.graph_id):
+            raise DGLError(
+                "More graph ids are found in node/edge data than graph data. Please specify all graph ids in graph data CSV.")
+        graph_ids = graph_data.graph_id
+        graphs = []
+        for graph_id in graph_ids:
+            if graph_id not in graphs_dict:
+                graphs_dict[graph_id] = dgl_heterograph(
+                    {('_V', '_E', '_V'): ([], [])})
+        for graph_id in graph_ids:
+            graphs.append(graphs_dict[graph_id])
+        data = {}
+        for k, v in graph_data.data.items():
+            data[k] = F.tensor(v)
```

Comment:
BTW, you could use list comprehension for dict as well: `data = {k : F.tensor(v) for k, v in graph_data.data.items()}`

Reply:
accepted.